File: LC100/lc85.js
function maxArea(matrix) {
    var result = 0;
    if (matrix === null || matrix.length === 0 || matrix[0].length === 0) {
        return result;
    }

    var nCol = matrix[0].length;
    var nRow = matrix.length;
    var height = Array(nCol+1).fill(0);
    var stack, height1, width;
    for (var row=0; row<nRow; row++) {
        for (var col=0; col<nCol; col++) {
            if (matrix[row][col] === "1") {
                height[col] += 1;
            } else {
                height[col] = 0;
            }
        }

        stack = [0];
        for (var i=1; i<nCol+1; i++) {
            while (stack.length>0 && height[i] < height[stack.slice(-1)[0]]) {
                height1 = height[stack.pop()];
                if (stack.length === 0) {
                    width = i;
                } else {
                    width = i - 1 - stack.slice(-1)[0];
                }
                result = Math.max(result, height1 * width);
            }
            stack.push(i);
        }
    }
    return result;
}


function test() {
    var matrixs = [
        [],
        ["10100", "10111", "11111", "10010"],
        ["110101", "010011", "111101", "111101"]
    ];
    for (var matrix of matrixs) {
        console.log(maxArea(matrix));
    }
}


test();

File: LC100/lc85.py
"""
Given a 2D binary matrix filled with 0's and 1's,
 find the largest rectangle containing only 1's and return its area.

For example, given the following matrix:
1 0 1 0 0
1 0 1 1 1
1 1 1 1 1
1 0 0 1 0
Return 6.
"""


def max_rectangle(matrix):
    result = 0
    if not matrix or len(matrix) == 0 or len(matrix[0]) == 0:
        return result

    n_col = len(matrix[0])
    n_row = len(matrix)
    height = [0 for _ in range(n_col + 1)]
    for row in range(n_row):
        for col in range(n_col):
            if matrix[row][col] == "1":
                height[col] += 1
            else:
                height[col] = 0

        stack = [0]
        for i in range(1, len(height)):
            while stack and height[i] < height[stack[-1]]:
                height1 = height[stack.pop()]
                width = i if not stack else i - stack[-1] - 1
                result = max(result, height1 * width)
            stack.append(i)
    return result


def test():
    matrix_list = [
        [""],
        ["10100", "10111", "11111", "10010"],
        ["110101", "010011", "111101", "111101"]
    ]

    for matrix in matrix_list:
        print(max_rectangle(matrix))


test()
